packages/contracts-bedrock/test/kontrol/README.md

```diff
@@ -0,0 +1,69 @@
+# Kontrol Verification
+
+This folder contains Kontrol symbolic property tests.
+
+## Directory structure
+
+The directory structure is as follows
+
+```tree
+test/kontrol
+├── Counter.sol
+├── Counter.t.sol
+├── kontrol
+│   ├── pausability-lemmas.k
+│   └── scripts
+│       ├── json
+│       │   ├── clean_json.py
+│       │   └── reverse_key_values.py
+│       ├── make-summary-deployment.sh
+│       ├── run-kontrol-local.sh
+│       └── run-kontrol.sh
+├── KontrolDeployment.sol
+└── proofs
+    ├── interfaces
+    │   └── KontrolInterfaces.sol
+    ├── OptimismPortal.k.sol
+    └── utils
+        ├── DeploymentSummaryCode.sol
+        ├── DeploymentSummary.sol
```

Comment:
These are autogenerated based on `KontrolDeployment.sol`? If so, we probably should not commit them and should add them to the `.gitignore`

Reply:
They are autogenerated. The point of keeping track of them would be to know the changes in the summary contracts due to either changes in your codebase or in Kontrol. It's not necessary to keep them there, though, since they'll be automatically generated. But if you want to not keep track of them, I'd still have them here while we polish this PR and remove them before merging

Reply:
The only time these should change is when the deploy script changes? If so, it's really serving as a snapshot, so we should either:
- Only track the resulting state diff JSON file that's used in `snapshots/state-diff/Deploy.json`
- Or, move these solidity files into `snapshots/deploy-script` (or similar).

The idea being that we have a guideline any changes to files in the `snapshot` folder should be carefully reviewed during a PR

Reply:
I've added the file `snapshots/statediff/Kontrol-Deploy.json` to keep track of the statediff. I tried removing the `DeploymentSummary*` files, but for some reason, the `--skip` feature is not working (or I don't know how to make it work properly), and the compilation stage will fail when trying to compile `OptimismPortal.k.sol`, which inherits from `DeploymentSummary.sol`.
As a workaround, I could try isolating `KontrolDeployment.sol` into its own folder and modifying the `kdeploy` profile not to attempt to compile the `proofs` dir.
Let me know if you have any ideas on this aspect!

Reply:
IIUC now any regular `forge test` will fail because of the missing deployment summary that we inherit from? If so one workaround is to instead do something like we do in `Setup.sol`:
```solidity
deploy = Deploy(_create(vm.getCode("Deploy.s.sol:Deploy")));
```

This uses a cheatcode to read the `Deploy` contract artifact and deploy it. We could do the same thing here instead of inheriting, that way we don't need the contract at build time, only compile time, and we can then call e.g. `deploymentSummary.apply()` to apply the state diff

Reply:
If I get it right, the change would be using cheatcodes to load the compiled code at runtime so that the file would compile well even without `DeploymentSummary.sol` being in place. Is this it?
If so, I see two main difficulties with this approach:
- The address variables in `DeploymentSummary` are `internal constant`s. This would mean we need to inherit directly from it to compile `OptimismPortal.k.sol` properly. We could have, as a workaround, the addresses be public. But this part of `DeploymentSummary`'s interface is automatically generated by the file in `deployments/hardhat/.deploy`, so that might be a point of flakiness
- We have to run `OptimismPortal.k.sol` in Kontrol. This means having to execute the `_create` logic (adding more computation overhead but might not be much) and implement the `vm.getCode` cheatcode (which might not be trivial), and execute it as well.

I think with the fast deployment summary (generating directly the KEVM initial state from the auto-generated deployment contract), the link between `DeploymentSummary` and `OptimismPortal.k.sol` will be weaker. But we will still need, for instance, an auto-generated contract with the correct address variables.

I don't know how dirty the following is, but it might work. We could leave placeholder comments in the `OptimismPortal.k.sol` that will be substituted with `awk` commands by the right thing when executing `run-kontrol.sh`, so that the file compiles correctly without the `DeploymentSummary` contract, but we make sure to add the correct code before executing it with Kontrol.
I'm thinking of transforming this code, which will be the default to build and compile
```solidity
// import { DeploymentSummary } from "./utils/DeploymentSummary.sol";

contract OptimismPortalKontrol is KontrolUtils/*, DeploymentSummary */ {

    address OptimismPortal = address(0);

    function test_proveWithdrawalTransaction_paused(....
```
into the appropriate code to symbolically execute the tests
```solidity
import { DeploymentSummary } from "./utils/DeploymentSummary.sol";

contract OptimismPortalKontrol is KontrolUtils, DeploymentSummary {

    function test_proveWithdrawalTransaction_paused(....
```
However, I'm not sure if this level of complexity/workaroundness is worth compared with having the generated solidity files being tracked. What do you think?

Reply:
Hmm so I'd definitely like to avoid as much hackiness and complexity as possible to ensure it's easy to understand and maintainable.

Am I correct that the two cleanest solutions are:
1. Ensuring we can use the forge `--skip` flag can be used to skip compiling all `*.k.sol` files, and then NOT commit `DeploymentSummary.sol`? But this would require updating our standard `forge build` commands to skip these files too.
2. Simply commit `DeploymentSummary.sol`?

if (2) is all that's needed to keep things simple, let's commit the file, and make sure that autogenerated files include a comment at the top explaining they are autogenerated. For example:

```
// This file was autogenerated by running `<command here>`. Do not edit this file manually.
```

Reply:
Yes, these two are the cleanest solutions I see. I'm also in favor of (2). I'll open an issue in Kontrol to add that comment at the top of the generated files!
What of `snapshots/statediff/Kontrol-Deploy.json`? Do we want to keep tracking it? Or now that we'll be tracking the autogenerated files that's enough?
Note: We'll have to commit 2 files, `DeploymentSummary.sol` and `DeploymentSummaryCode.sol`, both in the same folder

Reply:
[Issue created](https://github.com/runtimeverification/kontrol/issues/255) :slightly_smiling_face:

Reply:
Sounds great, thanks! I'd still like to have the file in the snapshots folder because the diff in that snapshot folder (when the deploy state diff changes) is what should inform PR reviewers that extra scrutiny is needed

Reply:
To resolve this:
- Possibly, with fast summarization, we won't need the auto-generated files, but we do in the meantime. So for this PR they will stay. There's still left to update Kontrol so that the informative message is auto-generated
- There is [`snapshots/state-diff/Kontrol-Deploy.json`](https://github.com/runtimeverification/_audits_Ethereum-optimism_pausability/blob/kontrol-proofs/packages/contracts-bedrock/snapshots/state-diff/Kontrol-Deploy.json) recording the state diffs for the modified Kontrol deployment

This conversation will be resolved once said comments are autogenerated

Reply:
Resolving this to merge the PR. To my knowledge, all issues/suggestions from this thread have been addressed. Let me know if this is not the case or if you think of something else to improve!